src/app/tests/TestReadInteraction.cpp

```diff
@@ -986,6 +999,88 @@ void TestReadInteraction::TestReadRoundtripWithSameDifferentPathsDataVersionFilt
     NL_TEST_ASSERT(apSuite, ctx.GetExchangeManager().GetNumActiveExchanges() == 0);
 }
 
+void TestReadInteraction::TestReadRoundtripWithEventStatusIBInEventReport(nlTestSuite * apSuite, void * apContext)
+{
+    TestContext & ctx = *static_cast<TestContext *>(apContext);
+    CHIP_ERROR err    = CHIP_NO_ERROR;
+
+    Messaging::ReliableMessageMgr * rm = ctx.GetExchangeManager().GetReliableMessageMgr();
+    // Shouldn't have anything in the retransmit table when starting the test.
+    NL_TEST_ASSERT(apSuite, rm->TestGetCountRetransTable() == 0);
+
+    GenerateEvents(apSuite, apContext);
+
+    auto * engine = chip::app::InteractionModelEngine::GetInstance();
+    err           = engine->Init(&ctx.GetExchangeManager());
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    EventManagement::GetInstance().SetBypassACL(chip::app::EventManagement::BypassACL::kAlwaysFail);
+
+    // When read events with concrete paths without enough privilege, we will return EventStatusIB
+    {
+        chip::app::EventPathParams eventPathParams[1];
+        eventPathParams[0].mEndpointId = kTestEndpointId;
+        eventPathParams[0].mClusterId  = kTestClusterId;
+        eventPathParams[0].mEventId    = kTestEventIdDebug;
+
+        ReadPrepareParams readPrepareParams(ctx.GetSessionBobToAlice());
+        readPrepareParams.mpEventPathParamsList    = eventPathParams;
+        readPrepareParams.mEventPathParamsListSize = 1;
+        readPrepareParams.mEventNumber             = 1;
+
+        MockInteractionModelApp delegate;
+        NL_TEST_ASSERT(apSuite, !delegate.mGotEventResponse);
+
+        app::ReadClient readClient(chip::app::InteractionModelEngine::GetInstance(), &ctx.GetExchangeManager(), delegate,
+                                   chip::app::ReadClient::InteractionType::Read);
+
+        err = readClient.SendRequest(readPrepareParams);
+        NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+        InteractionModelEngine::GetInstance()->GetReportingEngine().Run();
```

Comment:
Please don't make manual `Run()` calls.  Those should be happening automatically as long as we run the event loop (which we should).  If we make manual `Run` calls in tests, that just covers up bugs in Run scheduling.

Reply:
We don't have a function for one-shot event processing for now.

We have a comment for 
`This DOES NOT ensure that all pending events are serviced to completion (i.e timers, any ScheduleWork calls).`

In `DrainAndServiceIO`

But invoking works scheduled by `ScheduleWork` is what we want here.

This should be a follow up as "Support one-shot event dispatch for tests"

Reply:
> But invoking works scheduled by ScheduleWork is what we want here.

You may want to update to tip and read the updated documentation for DrainAndServiceIO.

Again, we should not be adding `Run` calls.